src/ffs/ffs.ts

```diff
@@ -0,0 +1,104 @@
+import pubTree, { PublicTree } from './pub/tree'
+import privTree, { PrivateTree } from './priv/tree'
+import pubNode, { PublicNode } from './pub/node'
+import { Tree, Node, Link } from './types'
+import { CID, FileContent } from '../ipfs'
+import pathUtil from './path'
+
+export class FFS {
+
+  root: PublicNode
+  pubTree: PublicTree
+  privTree: PrivateTree
+
+  constructor(root: PublicNode, pubTree: PublicTree, privTree: PrivateTree) {
+    this.root = root
+    this.pubTree = pubTree
+    this.privTree = privTree
+  }
+
+  async put(): Promise<CID> {
+    return this.root.put()
+  }
+
+  async cid(): Promise<CID> {
+    return this.put()
+  }
+
+  async listDir(path: string): Promise<Link[] | null> {
+    const { tree, tail } = this.parsePath(path)
+    return tree.listDir(tail)
+  }
+
+  async makeDir(path: string): Promise<FFS> {
+    const { tree, tail } = this.parsePath(path)
+    await tree.makeDir(tail)
+    return this.updateRoot()
+  }
+
+  async addFile(path: string, content: FileContent): Promise<FFS> {
+    const { tree, tail } = this.parsePath(path)
+    await tree.addFile(tail, content)
+    return this.updateRoot()
+  }
+
+  async getFile(path: string): Promise<FileContent | null> {
+    const { tree, tail } = this.parsePath(path)
+    return tree.getFile(tail)
+  }
+
+  async getNode(path: string): Promise<Node | null> {
+    const { tree, tail } = this.parsePath(path)
+    return tree.getNode(tail)
+  }
+
+  async updateRoot(): Promise<FFS> {
+    const pubCID = await this.pubTree.put()
+    const privCID = await this.privTree.put()
+    const pubLink = { name: 'public', cid: pubCID }
+    const privLink = { name: 'private', cid: privCID }
+    this.root.replaceLink(pubLink)
+    this.root.replaceLink(privLink)
+    return this
+  }
+
+  parsePath(path: string): { tree: Tree, tail: string } {
+    const parts = pathUtil.split(path)
+    const head = parts[0]
+    const tail = pathUtil.join(parts.slice(1))
+    let tree: Tree
+    if(head === 'public') {
+      tree = this.pubTree
+    }else if(head === 'private') {
+      tree = this.privTree
+    }else {
+      throw new Error("Not a valid FFS path")
+    }
+    return { tree, tail }
+  }
+}
+
+export async function empty(): Promise<FFS> {
+  const root = await pubNode.empty()
+  const pubTreeInstance = await pubTree.empty()
+  const privTreeInstance = await privTree.empty()
+  return new FFS(root, pubTreeInstance, privTreeInstance)
+}
+
+export async function resolve(cid: CID, keyStr: string): Promise<FFS | null> {
+  const root = await pubNode.resolve(cid)
+  const pubLink = root.findLink('public')
+  const privLink = root.findLink('private')
+  if(pubLink === null || privLink === null) {
+    return null
+  }
+  const pubTreeInstance = await pubTree.resolve(pubLink.cid)
+  const privTreeInstance = await privTree.resolve(privLink.cid, keyStr)
+  return new FFS(root, pubTreeInstance, privTreeInstance)
+}
+
+export default {
+  FFS,
```

Comment:
Do we need to export FFS at all? 🤷‍♂ 
Are there any use-cases besides `empty` and `resolve` that we have here?

Reply:
Yeah you might want to bootstrap an FFS off of an existing public folder so do something like
```ts
const root = await pubNode.empty()
const pubTreeInstance = await pubTree.resolve(PUB_TREE_CID)
const privTreeInstance = await privTree.empty()
const ffs = new FFS(root, pubTreeInstance, privTreeInstance)
}
```
Then again, you could also do
```ts
const ffs = await empty()
ffs.pubTree = await pubTree.resolve(PUBLIC_TREE_CID)
await ffs.updateRoot()
```
I was arguable over-aggressive with exports, but was trying to give as much flexibility as possible :laughing:

Reply:
Ah awesome, thanks for explaining 👍